tsl/src/chunk.c

```diff
@@ -536,25 +703,323 @@ chunk_copy_data(PG_FUNCTION_ARGS)
+		.scandirection = ForwardScanDirection,
+	};
+
+	ScanKeyInit(&scankey[0],
+				Anum_chunk_move_activity_pkey_idx_id,
+				BTEqualStrategyNumber,
+				F_INT4EQ,
+				Int32GetDatum(id));
+
+	return ts_scanner_scan(&scanctx);
+}
+
+static void
+chunk_move_activity_insert_rel(Relation rel, ChunkMoveData *ccd)
+{
+	TupleDesc desc = RelationGetDescr(rel);
+	Datum values[Natts_chunk_move_activity];
+	bool nulls[Natts_chunk_move_activity] = { false };
+	CatalogSecurityContext sec_ctx;
+
+	memset(values, 0, sizeof(values));
+	values[AttrNumberGetAttrOffset(Anum_chunk_move_activity_id)] = ccd->id;
+	values[AttrNumberGetAttrOffset(Anum_chunk_move_activity_operation_id)] =
+		NameGetDatum(&ccd->operation_id);
+	values[AttrNumberGetAttrOffset(Anum_chunk_move_activity_backend_pid)] = MyProcPid;
+	values[AttrNumberGetAttrOffset(Anum_chunk_move_activity_stage)] =
+		CStringGetTextDatum(ccd->stage);
+	values[AttrNumberGetAttrOffset(Anum_chunk_move_activity_time_end)] =
+		TimestampGetDatum(DT_NOBEGIN);
+	values[AttrNumberGetAttrOffset(Anum_chunk_move_activity_time_start)] =
+		TimestampGetDatum(GetCurrentTimestamp());
+	values[AttrNumberGetAttrOffset(Anum_chunk_move_activity_time_prev_stage_end)] =
+		TimestampGetDatum(DT_NOBEGIN);
+	values[AttrNumberGetAttrOffset(Anum_chunk_move_activity_chunk_id)] = ccd->chunk->fd.id;
+	values[AttrNumberGetAttrOffset(Anum_chunk_move_activity_source_node_name)] =
+		DirectFunctionCall1(namein, CStringGetDatum(ccd->src_node));
+	values[AttrNumberGetAttrOffset(Anum_chunk_move_activity_dest_node_name)] =
+		DirectFunctionCall1(namein, CStringGetDatum(ccd->dst_node));
+	values[AttrNumberGetAttrOffset(Anum_chunk_move_activity_is_move)] = BoolGetDatum(ccd->is_move);
+
+	/* nulls is FALSE above, insert the row now */
+	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+	ts_catalog_insert_values(rel, desc, values, nulls);
+	ts_catalog_restore_user(&sec_ctx);
+}
+
+static void
+chunk_move_activity_insert(ChunkMoveData *ccd)
+{
+	Catalog *catalog;
+	Relation rel;
+
+	catalog = ts_catalog_get();
+	rel = table_open(catalog_get_table_id(catalog, CHUNK_MOVE_ACTIVITY), RowExclusiveLock);
+
+	chunk_move_activity_insert_rel(rel, ccd);
+	table_close(rel, RowExclusiveLock);
+}
+
+static ScanTupleResult
+chunk_move_activity_tuple_update(TupleInfo *ti, void *data)
+{
+	ChunkMoveData *ccd = data;
+	Datum values[Natts_chunk_move_activity];
+	bool nulls[Natts_chunk_move_activity];
+	CatalogSecurityContext sec_ctx;
+	bool should_free;
+	HeapTuple tuple = ts_scanner_fetch_heap_tuple(ti, false, &should_free);
+	HeapTuple new_tuple;
+
+	heap_deform_tuple(tuple, ts_scanner_get_tupledesc(ti), values, nulls);
+
+	/* We only update the "stage" and time related fields */
+	values[AttrNumberGetAttrOffset(Anum_chunk_move_activity_stage)] =
+		CStringGetTextDatum(ccd->stage);
+	values[AttrNumberGetAttrOffset(Anum_chunk_move_activity_time_prev_stage_end)] =
+		TimestampGetDatum(GetCurrentTimestamp());
```

Comment:
```suggestion
	values[AttrNumberGetAttrOffset(Anum_chunk_move_activity_time_prev_stage_end)] =
		TimestampTzGetDatum(GetCurrentTimestamp());
```

GetCurrentTimeStamp() return the current OS time, not transaction time. Is this what we want? I guess it is, since otherwise the time returned would be the start of the transaction, right?

Reply:
Remvoed this column now.